nnet_utils/nnet_batchnorm.h

```diff
@@ -0,0 +1,139 @@
+    static const bool store_weights_in_bram = false;
+    static const unsigned n_zeros = 0;
+    // partitioning arrays cyclically to go with roll factors?
+};
+
+template<class data_T, class res_T, typename CONFIG_T>
+void normalize(
+    data_T    data[CONFIG_T::n_in],
+    res_T     res[CONFIG_T::n_out],
+    typename CONFIG_T::scale_t  scale[CONFIG_T::n_in],
+    typename CONFIG_T::beta_t   beta[CONFIG_T::n_in],
+    typename CONFIG_T::mean_t   mean[CONFIG_T::n_in])
+{
+    data_T cache;
+    typename CONFIG_T::accum_t mult[CONFIG_T::n_in];
+    typename CONFIG_T::accum_t acc[CONFIG_T::n_in];
+    typename CONFIG_T::accum_t shift[CONFIG_T::n_in];
+
+    
+    // Use a function_instantiate in case it helps to explicitly optimize unchanging weights/biases
+    #pragma HLS function_instantiate variable=scale,beta,mean
+
+    if (CONFIG_T::io_type == io_parallel){
+        // For parallel inputs:
+        //   - completely partition arrays -- target fabric
+        //   - if we have an unroll factor, limit number of multipliers
+        #pragma HLS PIPELINE II=CONFIG_T::reuse_factor
+
+        // #pragma HLS ARRAY_PARTITION variable=weights complete // remove this line for now, it breaks compression sometimes
+	#pragma HLS ARRAY_PARTITION variable=scale complete
+        #pragma HLS ARRAY_PARTITION variable=beta complete
+	#pragma HLS ARRAY_PARTITION variable=mean complete
+        #pragma HLS ARRAY_PARTITION variable=mult complete
+        #pragma HLS ARRAY_PARTITION variable=acc complete
+        #pragma HLS ARRAY_PARTITION variable=shift complete
+
+        int multiplier_limit  = ceil(float(CONFIG_T::n_in*CONFIG_T::n_out) / float(CONFIG_T::reuse_factor)) - floor(float(CONFIG_T::n_zeros) / float(CONFIG_T::reuse_factor));
+        #pragma HLS ALLOCATION instances=mul limit=multiplier_limit operation
+
+    } else if (CONFIG_T::io_type == io_serial){
+        #pragma HLS ARRAY_RESHAPE variable=scale complete dim=2
+        #pragma HLS ARRAY_RESHAPE variable=beta complete dim=2
+        #pragma HLS ARRAY_RESHAPE variable=mean complete dim=2
+        #pragma HLS ARRAY_PARTITION variable=mult complete dim=2
+        #pragma HLS ARRAY_PARTITION variable=acc complete dim=1
+        #pragma HLS ARRAY_PARTITION variable=shift complete dim=1
+        #pragma HLS DATAFLOW
+        #pragma HLS STREAM variable=mult depth=1
+        #pragma HLS STREAM variable=acc depth=1
+    }
+
+    // Shift input data by the mean
+    ShiftInputs: for(int i = 0; i < CONFIG_T::n_in; i++) {
+        if (CONFIG_T::io_type == io_serial){
+            #pragma HLS UNROLL
+        }
+        shift[i] = (typename CONFIG_T::accum_t) (data[i]-mean[i]);
+    }
+        
+    // Do the multiplication
+    Product: for(int i = 0; i < CONFIG_T::n_in; i++) {
+        if (CONFIG_T::io_type == io_serial){
+            #pragma HLS PIPELINE
+        }
+        mult[i] = scale[i] * shift[i];
+    }    
+    
+    // Initialize accumulator with beta values
```

Comment:
This code looks good, but i wouldn't really call it an accumulator because each entry is just taking one multiplication result.

Do you get different resource usage if you collapse all of these loops into just one loop with `output = (data-mean)*scale + beta`?

Reply:
I changed to calculate the result in one line as you suggested. It does not seem to affect latency nor resources.